practice_project/weather_forecast.py

```diff
@@ -0,0 +1,86 @@
+import requests
+from abc import ABC, abstractmethod
+
+class WeatherDataProvider(ABC):
+    @abstractmethod
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Abstract method to fetch weather forecast for a given latitude and longitude.
+        """
+        pass
+      
+class WeatherAPIDataProvider(WeatherDataProvider):
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Fetch weather forecast using Open-Meteo API.
+        """
+        api_url = f"https://api.open-meteo.com/v1/forecast?latitude={latitude}&longitude={longitude}&hourly=temperature_2m"
+        response = requests.get(api_url)
+        if response.status_code == 200:
+            return response.json()
+        else:
+            raise Exception("Failed to fetch weather forecast from API")
+          
+class MockWeatherDataProvider(WeatherDataProvider):
+    def fetch_weather_forecast(self, latitude, longitude):
+        """
+        Mocked method to return a dummy weather forecast.
+        """
+        return {
+            "latitude": latitude,
+            "longitude": longitude,
+            "hourly": {}
```

Comment:
here you should return some non-empty data, since other wise you don't really mock anything. You can check the recording from session#4 for an example